Buffer reuse

fegies opened the pull request:
Avoid reallocating the Protocol message buffer by introducing a Trait that abstracts over the concrete stream type and accepts a mutable buffer.

Furthermore, the ByteConsumer Iterator impl is now based on the FramedRead.

Will update the PR with profiling data once I am back at work with access to the test dataset

There are some open questions:
- Does it make sense to update the Converter struct to use the new Trait also?
- Should the minor version of the stream-delimit subcrate be bumped? (there should be strictly new features)

EDIT: 

Time after this PR:
$ time ./pq --msgtype simmo.TranshipmentCandidate --fdsetdir ./fdset --stream i32be < candidates.proto.ld > /dev/null
real    1m12.450s
user    1m11.133s
sys     0m1.047s

For comparison: achieved Result from previous PR:
$ time ./pq --msgtype simmo.TranshipmentCandidate --fdsetdir ./fdset --stream i32be < candidates.proto.ld > /dev/null
real 1m19.370s
user 1m14.457s
sys 0m4.568s

fegies commented on stream-delimit/src/byte_consumer.rs:
```diff
@@ -16,30 +19,65 @@ impl<T: Read> ByteConsumer<T> {
+                    match e {
+                        StreamDelimitError::VarintDecodeError(i) => i,
+                        e => io::Error::new(io::ErrorKind::InvalidData, format!("{}", e)),
+                    }
```
Not sure if its fine to just shunt all errors into the InvalidData kind

fegies commented:
Here is a flamegraph after:
![flame_bufferreuse](https://user-images.githubusercontent.com/20617897/102077460-a3488a00-3e09-11eb-885a-1091d38d1c29.PNG)

And here before (with the saved part crudely marked)
![prebuffer](https://user-images.githubusercontent.com/20617897/102078147-ba3bac00-3e0a-11eb-9d17-4f868b41fadc.PNG)


Looking at the graphs, it would seem there are still significant savings to be had in the serde_protobuf library, though I am not sure if this is feasible with justifiable effort

sevagh commented:
>Does it make sense to update the Converter struct to use the new Trait also?

I'll have to review this PR more closely so I can have an answer for this. It could take more time.

>Should the minor version of the stream-delimit subcrate be bumped? (there should be strictly new features)

Unless you really want to, you don't need to worry about version bumps - after merging all the PRs, I can take care of the version bumps for the next release.

fegies commented:
So, I had a thorough look at serde-protobuf, and found some unfortunate things:
- because of the protobuf spec, the reader should only keep the last entry for a given key. This requires at least buffering the entire message or deserializing and overwriting duplicate subobjects as the parser descends the structure.
- serde-protobuf chooses the latter, leading to a tree of keys being constructed internally for every message and submessage
- because the parser structure does not have any lifetime references to the parsed data, its not possible to buffer the message and descend lazily, or reuse sets or maps without major effort.
- this is complicated because it uses the `protobuf` crate, with the CodedInputStream as an abstration over a BufferedStream -> no zero-copy here possible either.

On the pq-side, I did not find any area where perf could be easily improved. (aside from changing the protobuf library).

So, aside from addressing comments from your side, I'd consider this PR series finished.

Cheers!

sevagh commented:
Verified on my end with pmap and the new PR has much less memory used. Tests passing. Thanks for another great contribution.
